src/scene/vcPolyModelNode.cpp

```diff
@@ -0,0 +1,92 @@
+#include "vcPolyModelNode.h"
+
+#include "vcPolygonModel.h"
+#include "vcStrings.h"
+#include "vcRender.h"
+
+#include "udPlatformUtil.h"
+#include "udStringUtil.h"
+
+#include "imgui.h"
+#include "imgui_ex/vcImGuiSimpleWidgets.h"
+
+vcPolyModelNode::vcPolyModelNode(vdkProject *pProject, vdkProjectNode *pNode, vcState *pProgramState) :
+  vcSceneItem(pProject, pNode, pProgramState)
+{
+  m_pModel = nullptr;
+  m_matrix = udDouble4x4::identity();
+
+  //TODO: Do this load async
+  if (vcPolygonModel_CreateFromURL(&m_pModel, pNode->pURI) == udR_Success)
+    m_loadStatus = vcSLS_Loaded;
+  else
+    m_loadStatus = vcSLS_Failed;
+}
+
+void vcPolyModelNode::OnNodeUpdate(vcState * /*pProgramState*/)
+{
+  // Does nothing
+}
+
+void vcPolyModelNode::AddToScene(vcState * /*pProgramState*/, vcRenderData *pRenderData)
+{
+  if (m_loadStatus != vcSLS_Loaded || !m_visible)
+    return;
+
+  vcRenderPolyInstance *pModel = pRenderData->polyModels.PushBack();
+  pModel->pModel = m_pModel;
+  pModel->pSceneItem = this;
+  pModel->worldMat = m_matrix;
+}
+
+void vcPolyModelNode::ApplyDelta(vcState * /*pProgramState*/, const udDouble4x4 &delta)
+{
+  m_matrix = delta * m_matrix;
+}
+
+void vcPolyModelNode::HandleImGui(vcState *pProgramState, size_t *pItemID)
+{
+  if (pProgramState->settings.presentation.showDiagnosticInfo)
+  {
+    for (int i = 0; i < m_pModel->meshCount; ++i)
+    {
+      ImGui::Text("%s %d / %s", vcString::Get("polyModelMesh"), i, m_pModel->pMeshes[i].material.pName);
+      ImGui::Indent();
+
+      if (m_pModel->pMeshes[i].material.pTexture != nullptr && m_pModel->pMeshes[i].material.pTexture != pProgramState->pWhiteTexture)
```

Comment:
This won't work for OBJ materials with no textures. They don't use `pProgramState->pWhiteTexture'. They internally create their own 1x1 white texture.